src/MagentoHackathon/Composer/Magento/Repository/InstalledPackageFileSystemRepository.php

```diff
@@ -145,20 +121,15 @@ public function add(InstalledPackage $package)
     {
         $this->load();
 
-        $this->hasChanges = true;
-
         try {
             $this->findByPackageName($package->getName());
-            foreach ($this->packages as &$installedPackage) {
-                if ($installedPackage->getName() === $package->getName()) {
-                    $installedPackage = $package;
-                }
-            }
         } catch (\Exception $e) {
             $this->packages[] = $package;
+            $this->hasChanges = true;
             return;
         }
 
+        throw new \Exception(sprintf('Package: "%s" is already installed', $package->getName()));
```

Comment:
It is not called twice. On updating the package is first removed from the repository. A package should not be added twice hence throwing an exception.